Fix StartTLS_Init

yota22721 opened the pull request:
Fixed `starttlsCmd` comparison inside the `StartTLS_Init` function

When using smtp STARTTLS, tmpBuf would be like `220 smtp.gmail.com ESMTP 13-20020a170902c24d00b00189348ab156sm5773120plg.283 - gsmtp`. `tmpBuf` won't match with `starttlsCmd`. I added memcpy to avoid the use of strncmp.

dgarske commented:
OK to test

dgarske reviewed (approved):
@yota22721 the changes look fine to convert these strcmp functions to safer strncmp. To be clear the startTLS was working correctly before. The goal here is to avoid the use of strncmp, correct?

yota22721 commented:
When using smtp STARTTLS, tmpBuf would be like ```220 smtp.gmail.com ESMTP 13-20020a170902c24d00b00189348ab156sm5773120plg.283 - gsmtp```. tmpBuf won't match with starttlsCmd, so I changed XSTRNCMP at the first commit. Now at https://github.com/wolfSSL/wolfssl/pull/5907/commits/67c6d73a8fa5506970251789c024c19639168be8,  
I added memcpy to avoid the use of strncmp.

dgarske commented on examples/client/client.c:
```diff
@@ -834,15 +834,17 @@ const char* starttlsCmd[6] = {
     if (recv(*sockfd, tmpBuf, sizeof(tmpBuf)-1, 0) < 0)
         err_sys("failed to read STARTTLS command\n");
 
+    XMEMCPY(buf,tmpBuf,XSTRLEN(starttlsCmd[0]));
```
Please use spaces after the comma's `, `. Shouldn't the copy include the null terminating character? So use `XSTRLEN() + 1`?

dgarske commented on examples/client/client.c:
```diff
@@ -834,15 +834,17 @@ const char* starttlsCmd[6] = {
 static int StartTLS_Init(SOCKET_T* sockfd)
 {
     char tmpBuf[512];
+    char buf[4];
```
Add trailing code comment. `/* 3 digit number plus 1 null char */`

dgarske commented on examples/client/client.c:
```diff
@@ -834,6 +834,7 @@ const char* starttlsCmd[6] = {
 static int StartTLS_Init(SOCKET_T* sockfd)
 {
     char tmpBuf[512];
+    char buf[3];     /* 3 digit number */
```
The `buf` should still have a NULL terminator. Please make `buf[4]`. The XMEMCPY needs to be `XSTRLEN()+1`.
Also you still need to add spaces after the comma's `, ` for function arguments.

douzzer commented on examples/client/client.c:
```diff
@@ -843,7 +844,9 @@ static int StartTLS_Init(SOCKET_T* sockfd)
         err_sys("failed to read STARTTLS command\n");
 
-    if (!XSTRCMP(tmpBuf, starttlsCmd[0])) {
+    XMEMCPY(buf, tmpBuf, XSTRLEN(starttlsCmd[0])+1);
```
Before each `XMEMCPY()`, you need to check that `XSTRLEN(tmpBuf) >= XSTRLEN(starttlsCmd[the number])+1`, and check that `tmpBuf[XSTRLEN(starttlsCmd[the number]] == ' '` to confirm that the input has a space character where expected.

In the alternative you can skip the extra `buf` and the `XMEMCPY()`s, and use the simpler approach I showed in my earlier comment.

dgarske commented on examples/client/client.c:
```diff
@@ -834,15 +834,17 @@ const char* starttlsCmd[6] = {
         err_sys("failed to read STARTTLS command\n");
 
+    XMEMCPY(buf,tmpBuf,XSTRLEN(starttlsCmd[0]));
     if (!XSTRCMP(tmpBuf, starttlsCmd[0])) {
```
@yota22721 please use this suggestion from douzzer for the fix. Thanks

douzzer reviewed (approved):
Looks good -- tested (after rebasing) with `wolfssl-multi-test.sh ... super-quick-check`, notably the sanitizers, `cppcheck`, and `clang-tidy`.

dgarske commented on examples/client/client.c:
```diff
@@ -843,7 +843,8 @@ static int StartTLS_Init(SOCKET_T* sockfd)
         err_sys("failed to read STARTTLS command\n");
 
-    if (!XSTRCMP(tmpBuf, starttlsCmd[0])) {
+    if ((!XSTRNCMP(tmpBuf, starttlsCmd[0], XSTRLEN(starttlsCmd[0])))&&
```
Please use space before `&&`. See for other two.

dgarske commented:
@yota22721 please quash the PR. It won't pass the FIPSv2 test because some of the commits conflict.

dgarske commented:
@yota22721 please update this PR with the steps used to test this change. Thank you

dgarske reviewed (approved):
Thanks Yota.
